File: day-7-facial-landmark-detection/facialLandmarkDetector.cpp
/*
Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED

This code is made available to the students of 
the online course titled "OpenCV for Beginners" 
by Satya Mallick for personal non-commercial use. 

Sharing this code is strictly prohibited without written
permission from Big Vision LLC. 

For licensing and other inquiries, please email 
spmallick@bigvisionllc.com 
*/


#include <dlib/opencv.h>
#include <dlib/image_processing.h>
#include <dlib/image_processing/frontal_face_detector.h>
#include <opencv2/opencv.hpp>

using namespace dlib;
using namespace std;
using namespace cv;

// Write landmarks to file
void writeLandmarksToFile(full_object_detection &landmarks, const string &filename)
{
  // Open file
	std::ofstream ofs;
	ofs.open(filename);
  
  // Loop over all landmark points
  for (int i = 0; i < landmarks.num_parts(); i++)
	{
    // Print x and y coordinates to file
		ofs << landmarks.part(i).x() << " " << landmarks.part(i).y() << endl;

	}
  // Close file
	ofs.close();
} 

void drawLandmarks(Mat &im, full_object_detection landmarks )
{
  for(int i = 0; i < landmarks.num_parts(); i++)
  {
    int px = landmarks.part(i).x();
    int py = landmarks.part(i).y();

    char landmark_index[3];
    sprintf(landmark_index, "%d", i+1);

    circle(im, Point(px, py), 1, Scalar(0, 0, 255), 2, CV_AA);
    putText(im, landmark_index, Point(px, py), FONT_HERSHEY_SIMPLEX, .3, Scalar(255, 0, 0), 1);
  }
}

int main(int argc, char** argv)
{  
  
  // Get the face detector
  frontal_face_detector faceDetector = get_frontal_face_detector();

  // The landmark detector is implemented in the shape_predictor class
  shape_predictor landmarkDetector;

  // Load the landmark model
  deserialize("shape_predictor_68_face_landmarks.dat") >> landmarkDetector;

  // Read Image
  string imageFilename("girl.jpg");
  cv::Mat im = cv::imread(imageFilename);
  
  // landmarks will be stored in results/famil_0.txt
  string landmarksBasename("output_cpp");
  
  // Convert OpenCV image format to Dlib's image format
  cv_image<bgr_pixel> dlibIm(im);

  // Detect faces in the image
  std::vector<dlib::rectangle> faceRects = faceDetector(dlibIm);
  cout << "Number of faces detected: " << faceRects.size() << endl;

  // Vector to store landmarks of all detected faces
  std::vector<full_object_detection> landmarksAll;

  // Loop over all detected face rectangles
  for (int i = 0; i < faceRects.size(); i++)
  {
    // For every face rectangle, run landmarkDetector
    full_object_detection landmarks = landmarkDetector(dlibIm, faceRects[i]);
    
    // Print number of landmarks
    if (i == 0) cout << "Number of landmarks : " << landmarks.num_parts() << endl;
    
    // Store landmarks for current face
    landmarksAll.push_back(landmarks);
    
    // Draw landmarks on face
    drawLandmarks(im, landmarks);

    // Write landmarks to disk
    std::stringstream landmarksFilename;
    landmarksFilename << landmarksBasename <<  "_"  << i << ".txt";
    cout << "Saving landmarks to " << landmarksFilename.str() << endl;
    writeLandmarksToFile(landmarks, landmarksFilename.str());
    
  }
  
  // Save image
  string outputFilename("result_cpp_Landmarks.jpg");
  cout << "Saving output image to " << outputFilename << endl;
  cv::imwrite(outputFilename, im);
  
  // Display image
  cv::imshow("Facial Landmark Detector", im);
  cv::waitKey(0);
  
  return EXIT_SUCCESS;
}

File: day-7-facial-landmark-detection/facialLandmarkDetector.py
# Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED
# 
# This code is made available to the students of 
# the online course titled "OpenCV for Beginners" 
# by Satya Mallick for personal non-commercial use. 
#
# Sharing this code is strictly prohibited without written
# permission from Big Vision LLC. 
#
# For licensing and other inquiries, please email 
# spmallick@bigvisionllc.com 
#


import dlib,cv2
import numpy as np

def writeLandmarksToFile(landmarks, landmarksFileName):
  with open(landmarksFileName, 'w') as f:
    for p in landmarks.parts():
      f.write("%s %s\n" %(int(p.x),int(p.y)))

  f.close()

def drawLandmarks(im, landmarks):
  for i, part in enumerate(landmarks.parts()):
    px = int(part.x)
    py = int(part.y)
    cv2.circle(im, (px, py), 1, (0, 0, 255), thickness=2, lineType=cv2.LINE_AA)
    cv2.putText(im, str(i+1), (px, py), cv2.FONT_HERSHEY_SIMPLEX, .3, (255, 0, 0), 1)

# Landmark model location
PREDICTOR_PATH = "shape_predictor_68_face_landmarks.dat"
# Get the face detector
faceDetector = dlib.get_frontal_face_detector()
# The landmark detector is implemented in the shape_predictor class
landmarkDetector = dlib.shape_predictor(PREDICTOR_PATH)

# Read image
imageFilename = "girl.jpg"
im= cv2.imread(imageFilename)
# landmarks will be stored in results/family_i.txt
landmarksBasename = "output_python"

# Detect faces in the image
faceRects = faceDetector(im, 0)
print("Number of faces detected: ",len(faceRects))

# List to store landmarks of all detected faces
landmarksAll = []

# Loop over all detected face rectangles
for i in range(0, len(faceRects)):
  newRect = dlib.rectangle(int(faceRects[i].left()),int(faceRects[i].top()),
      int(faceRects[i].right()),int(faceRects[i].bottom()))

  # For every face rectangle, run landmarkDetector
  landmarks = landmarkDetector(im, newRect)
  # Print number of landmarks
  if i==0:
    print("Number of landmarks",len(landmarks.parts()))

  # Store landmarks for current face
  landmarksAll.append(landmarks)
  # Draw landmarks on face
  drawLandmarks(im, landmarks)

  landmarksFileName = landmarksBasename +"_"+ str(i)+ ".txt"
  print("Saving landmarks to", landmarksFileName)
  # Write landmarks to disk
  writeLandmarksToFile(landmarks, landmarksFileName)

outputFileName = "result_python_Landmarks.jpg"
print("Saving output image to", outputFileName)
cv2.imwrite(outputFileName, im)

cv2.imshow("Facial Landmark detector", im)
cv2.waitKey(0)
cv2.destroyAllWindows()

  